src/main/java/io/reactivex/internal/subscribers/flowable/InnerQueuedSubscriber.java

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright 2016 Netflix, Inc.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the specific language governing permissions and limitations under the License.
+ */
+
+package io.reactivex.internal.subscribers.flowable;
+
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.reactivestreams.*;
+
+import io.reactivex.internal.fuseable.QueueSubscription;
+import io.reactivex.internal.subscriptions.SubscriptionHelper;
+import io.reactivex.internal.util.QueueDrainHelper;
+
+/**
+ * Subscriber that can fuse with the upstream and calls a support interface
+ * whenever an event is available.
+ *
+ * @param <T> the value type
+ */
+public final class InnerQueuedSubscriber<T> 
```

Comment:
Maybe add separate unit tests for it and check that behavior of public API is correct and it correctly reacts to different fusion modes? Up to you of course